### ISSUE TITLE
decorator does not satisfy ReportAbstractUsage's analysis of a Protocol; neither does in-line function containing definition

### ISSUE THREAD

**Issue by syncosmic** [microsoft/pyright#11688]
**The Bug**

A decorator which satisfies a `Protocol` by injecting an assignment into an `__init__` method doesn't get identified as implementing the `Protocol`, so `reportAbstractUsage` still considers the class abstract and says it cannot be instantiated (although it can).

In the toy example below, the `@with_logger` method drops `log_func` into a method of a class that follows the `HasLogFunction` protocol. `Test1` uses the decorator in its `__init__` whereas `Test2` does the same thing manually. Everything works fine except that Pyright gives `Test1` the same error as `TestE`, which _does_ fail to implement the function.

Obviously, following the decorator is a harder task for the static analyzer, so there may be practical limit here, but the error on `Test1` does seem like an incorrect conclusion.

**Quick Comparisons**

`mypy` seems to have the same behavior as Pyright (errors on `Test1` and `TestE`: `Cannot instantiate abstract class with abstract attribute "Log"  [abstract]`).

`ty` doesn't report an error on `Test1`, but don't get too jealous, because it doesn't report an error on `TestE` either, leaving a very obvious problem entirely up to runtime. Abstract class checking may be on its feature roadmap--I'm not following `ty` development (yet).

**MRE (or at least the simplest example I could come up with)**

[Pyright Playground](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBAE2AZwFCoGZQiOGCuAdgMYwQQA2ycokscA7lAIZhpY5zGUUCmpwCIWQA6ZgCNiNcNHgBhZhQoS%2BmbLhgBPMMEIBzaXXgBBQpoA0XZshiWACtjLcK6BLwxwKEPQH0CJAAoADwAuGygASjgAWgA%2BOAA5IV4Q1Dh0uDAoXRhgiPRiZWRqAAlrABlvADEiASEAhwgnSgjUjLhKvRCFJRVeAG1%2B8IBdSyTCXmHXd0ZgGAALHy89PV4ofqqQuB7lcT5%2BkUPLU01h4YD-Yi2qqLi4TbSMgAEmVmQL2vz2tw8GOcXlqsoD5dBMoAFdDYtmVkJ0aiQYIJCJYAFTMKB6ZBbCgoGD9E6jOAolEAawY6MxWyQpHxZmOZmGrUe7XSkJgIk6cAAvJ5vH5asyWVBeDB8FBCHhahDhLYiRTkKjSeSMcgvhlhaLxVYbAEqpY-gslt4gSDCGD8qhCtZqAAVXg2ACMARhcNqiKETPaL3%2BRpWa0FPzgPlNc2DAWQvAoGE9LPSYGt6G%2BM2WPhszFF7wjUZjsazGA53gCACI0xm4AB5ADSAEIixarcU4HabAAmZ0VapupE5xAzYO6UM%2BcOR6NtWNwPMFgw8lOXRMZQMp0v4TMjnvtSedYvL6hV2v1oq2%2B0wACi7dhnYR3bH6UD-cIg%2BH2ZvLPjxXnt%2BTfJ3T9Hgo3I5TtuMDpiuFY1nW6A2g63JNseTpRAAxHAwpGMY4jhMwpAAKrIMwqxbAoZpNDQMrMIQiLprwcASJhpBcIeqA2i2sHNjAbZIXAASEBAcBrNgkRMSerHHmenGobI6F0TAuH4Sk2zkTx8BsuRlEwNRtEwCw9ENmgTEOiIS4gRmAScQw0AktQGC6LwTEtoZ37GSuplwMh5lQJZeA2egWmaC%2BNong5vg-uJRCIiA1H8dAqC8EExC8GAJgwFpwDiPg6knlAAkvlkOTFsQzD4Ho8zwOmKVpepfFZdApFNiekGoMh0E0YQCBNixKkkPaNHChOWn4KQYqKBQmikUgNi6Ho%2BAoPMfSoNaazwEg4LQVEXI8stATMRayHkW1Cy8KNK7UQdE7MBFvIGJc7qEPNxSLU2Bmcig7VTugTUwXt7WkXM1DotRSAYBgay8BRXGnW43AsGQUAobwACO03CtQyA4NRujLfwN35LlFEBAA5AAOoQ63rY9QYhjAwb4-k40iONW0Gfeg449keNEyTpPfczVM%2BDTqB0wzzEiDzwY7XAhwiBVcCnUZoFWV2Qh-b14iaOp3BuDQbgUcABUuLjuQc6TPLNXLGb84LKCM0FqZOaq6AGwTxPG99ZsrhbKD01bwtu6qQA)

```python
from functools import wraps
from collections.abc import Callable
from typing import Any, cast, Protocol

def log_func(x:str) -> None:
    print(x)

class HasLogFunction(Protocol):
    Log:Callable[[str], None]

def with_logger[F: Callable[..., Any]](func: F) -> F:
    @wraps(func)
    def with_logger_inner(inst: HasLogFunction, *args: list[Any], **kwargs: dict[Any, Any]):
        inst.Log = log_func
        return func(inst, *args, **kwargs)
    return cast(F, with_logger_inner)

class Test1(HasLogFunction):
    @with_logger
    def __init__(self):
        pass

    def log_status(self):
        self.Log("status OK!")

class Test2(HasLogFunction):
    def __init__(self):
        self.Log = log_func

    def log_status(self):
        self.Log("status OK!")

class TestE(HasLogFunction):
    def __init__(self):
        pass

    def log_status(self):
        self.Log("status OK!")

T1 = Test1() # reportAbstractUsage: Cannot instantiate abstract class
T2 = Test2() # (no error)
TE = TestE() # reportAbstractUsage: Cannot instantiate abstract class

T1.log_status() # works fine
T2.log_status() # works fine

try:
    TE.log_status() # runtime error
except AttributeError:
    print("caught attribute error in TE")
```

The Pyright Playground link contains some additional output showing that the instances (T1 and T2) are structurally identical and that their log_status() methods are bytecode-identical.

**Context**
I'm using current Pyright in VS Code, but the result is the same in Pyright Playground (pyright 1.1.413 strict [which doesn't matter to this issue], python 3.14).

**Related Issues**
I couldn't find much on point (apologies if I missed anything); there is probably some overlap with [11411:open](https://github.com/microsoft/pyright/issues/11411) but that concerns a decorator on the class.

**Thanks**
As always! Pyright is great.

**Comment by syncosmic** [comment]
Turns out this is not specific to decorators. [Updated playground code](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBAE2AZwFCoGZQiOGCuAdgMYwQQA2ycokscA7lAIZhpY5zGUUCmpwCIWQA6ZgCNiNcNHgBhZhQoS%2BmbLhgBPMMEIBzaXXgBBQpoA0XZshiWACtjLcK6BLwxwKEPQH0CJAAoADwAuGygASjgAWgA%2BOAA5IV4Q1Dh0uDAoXRhgiPRiZWRqAAlrABlvADEiASEAhwgnSgjUjLhKvRCFJRVeAG1%2B8IBdSyTCXmHXd0ZgGAALHy89PV4ofqqQuB7lcT5%2BkUPLU01h4YD-Yi2qqLi4TbSMgAEmVmQL2vz2tw8GOcXlqsoD5dBMoAFdDYtmVkJ0aiQYIJCJYAFTMKB6ZBbCgoGD9E6jOAolEAawY6MxWyQpHxZmOZmGrUe7XSkJgIk6cAAvJ5vH5asyWVBeDB8FBCHhahDhLYiRTkKjSeSMcgvhlhaLxVYbAEqpY-gslt4gSDCGD8qhCtZqAAVXg2AAMARhcNqiKETO%2BMx8prmPoCyF4FAwnpZ6R%2BcGYCAQRt8lwCobD7UDwY53m5vLjtTgAGI4LwSF5kLoDDy9F5xIp0EnI9HY-zAhavR5lj4bMxRe8UyG2knu2m9AEAETtztwADyAGkAIRDi1W4pwO02ACMzoq1TdSMTcBe-1jQMFEZ9uj9PgDQZ7gvaYGt1YyEdbo-wXcvO%2BTl4Hw%2Bf1Cns-nRS2vaMAAEzrrCm4Ituvbht6vowP63bvhk-acjyraXPesEtnyP4XsGyHpKh3jfjAHYvhOM5zgUgFLsBACi4GulBHowYgcGngh55IWxN53keMxPmRnb4VeNbEYOI7CRRf7UagNoOhmy4wE6UR5sKRjGOI4TMKQACqyDMKsWwKGaTQ0DKzCEIiHa8JG2kwCwpBcIB8krkpwFrmpcAabIWk6fphnGdsVmEOZbJWTZMB2RIAXwAuaA2iBHk2GB3kBGF%2BZQNgkTyfRKUwIx3m%2BbA-mObpMAGUZKQhWZ8ARdZwC2fZcUufxCkiEJ5HvN5DDQCS1AYLovBuZ1uHST1uaMP1g3DfJIFjb4eG9TNeBzagjmaGxNr0YtbYTQmU1QEQiIgHZaw5agvBBMQvBgCYMCOcA4j4NF9HZdAbFZDkw7EMw%2BB6PM8Adk9L3RVlOUWUu9FyXmCmWDa7lWQgS7JQ1t3UOidnhPgpBiooFCaBZSA2CW%2BAoPMfSoNaazwEg4IKVEXI8vTASI-kNP0Kz7PcizwAMyBFp5sjcALLwRMvnZYtwIZZ2ZpKLGENTxS00uDoDjQtorgOyuBvQiMaygqM66gcMOgjSOECjSUWXMmPCogwAYBgawFvAATS6CawK3UEpGxA4gAG6CC%2BADcovzHZbjcCwZBQD5vAAI7k8K1DIDgdm6PT-DuoQ%2BTfdZAQAOQADqEMzzNq3AJ6EGeRf5CTIgk2z6s12e%2BfZIXpflxXS7uW3XH16gjfNwbA8%2Bh3P3dxXPI2%2BPPhDyPKBswt89C3AhwiGDEd2V1Y6XLn9t2eImjRdwbg0G4jV-S4Be5NPvcKZm%2B3dYvKBN8vHV7y%2BFp38XZczz7s-H8b9RCj21t-VU6A-4P0rjbSBoCP7vCSntH8EQgA) adds `Test0` (see below), which does what the decorator does, but in-line, and that also doesn't get marked as implementing the Protocol:

```python
class Test0(HasLogFunction):
    def __init__(self):
        def add_log_func():
            self.Log = log_func # enclosing = global

        add_log_func()

    def log_status(self):
        self.Log("status OK!")

T0 = Test0() # reportAbstractUsage: Cannot instantiate abstract class
```

I'm aware that this may straddle the line between a bug report and a feature request. The Pyright code (haven't looked yet) may well just look for direct assignments in `__init__`, which would probably catch most implementations, but I read the [typing spec for Protocol](https://typing.python.org/en/latest/spec/protocol.html) as broader:

> If a class defines all attributes and methods of a protocol with types that are assignable to the types of the protocol’s attributes and methods, it is said to implement the protocol and to be assignable to the protocol.

I suppose my contention is that Test0 (inner function) and Test1 (decorator) define `self.Log` just as much as Test2 (direct assignment) does, in a way that is exactly outcome-equivalent, so if it's feasible to update the static analyzer to catch at least these assignments to `self` by running through the call tree, that would be both more conformant and a nice feature. The decorator-satisfying-a-Protocol pattern, at least, is actually quite useful :)

**Comment by bschnurr** [comment]
Pyright does not perform interprocedural analysis through arbitrary function calls or decorators to determine which instance variables they assign. Doing so would require executing or interpreting general Python code and would not be reliable.

Explicitly inheriting from a protocol makes the class subject to abstract protocol-member checks. If the class is intended to satisfy the protocol structurally, it generally should not inherit from the protocol directly. Alternatively, declare or initialize `Log` explicitly in the concrete class.

The same limitation explains the nested-function variation. Pyright recognizes direct assignments in `__init__`, but it does not trace calls from `__init__` to discover indirect assignments.

Given these constraints, this is working as designed.

**Comment by syncosmic** [comment]
Got it. Thanks for your time!

Philosophically, I think that a lot of useful cases could be very reliably caught with some simple interprocedural analysis, but I (now) understand that it's a huge FR that's probably best suited to some hypothetical other product. I would note (without criticism) that Pyright is already "unreliable" in the other direction, since the below will pass the protocol member checks and then wait for a run-time error:

```python
class Test2(HasLogFunction):
    def __init__(self):
        self.Log = log_func
        self.__dict__.pop('Log')

    def log_status(self):
        self.Log("status OK!")
```

I appreciate the thought on protocol inheritance; I fear that I'm using the protocol /decorator combination in the best way currently available given my (actual, more complicated) use case, but I'll take another look and see if there's a better factoring available. In any case, I have a way to route around the limitation, it's just unaesthetic.

Thanks again!
